collector/zfs.go

```diff
@@ -0,0 +1,157 @@
+package collector
+
+// +build !nofilesystem
+// +build !nozfs
+
+import (
+	"errors"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+)
+
+type zfsMetricValue int
+
+const zfsErrorValue = zfsMetricValue(-1)
+
+var zfsNotAvailableError = errors.New("ZFS / ZFS statistics are not available")
+
+type zfsSysctl string
+type zfsSubsystemName string
+
+const (
+	arc = zfsSubsystemName("zfs_arc")
+)
+
+//------------------------------------------------------------------------------
+//                                    Metrics
+//------------------------------------------------------------------------------
+
+type zfsMetric struct {
+	subsystem zfsSubsystemName // The Prometheus subsystem name
+	name      string           // The Prometheus name of the metric
+	sysctl    zfsSysctl        // The sysctl of the ZFS metric
+}
+
+func NewZFSMetric(subsystem zfsSubsystemName, sysctl, name string) zfsMetric {
+	return zfsMetric{
+		sysctl:    zfsSysctl(sysctl),
+		subsystem: subsystem,
+		name:      name,
+	}
+}
+func (m *zfsMetric) BuildFQName() string {
+	return prometheus.BuildFQName(Namespace, string(m.subsystem), m.name)
+}
+
+func (m *zfsMetric) HelpString() string {
+	return m.name
+}
+
+//------------------------------------------------------------------------------
+//                                 Collector
+//------------------------------------------------------------------------------
+
+func init() {
+	Factories["zfs"] = NewZFSCollector
+}
+
+type zfsCollector struct {
+	zfsMetrics     []zfsMetric
+	metricProvider zfsMetricProvider
+}
+
+func NewZFSCollector() (Collector, error) {
+	err := zfsInitialize()
+	switch {
+	case err == zfsNotAvailableError:
+		log.Debug(err)
+		break
+		return &zfsCollector{}, err
+	}
+
+	return &zfsCollector{
+		zfsMetrics: []zfsMetric{
+			NewZFSMetric(arc, "kstat.zfs.misc.arcstats.p", "mru_size"),
+			NewZFSMetric(arc, "kstat.zfs.misc.arcstats.size", "size"),
```

Comment:
Not sure if I get you right.
Are you worried about exposing `c_min` as `target_min_size`? This is not specific to Linux.

Are you worried about Linux using different sysctls than FreeBSD / other ZFSs one day?
FreeBSD and Linux both use OpenZFS so there's a common code base.
The sysctls merely serve as an identifier for the metrics, which made sense at the time the exporter was FreeBSD-only.
AFAIK, the sysctls have been brought over from the Sun-Implementation and didn't change since.
I don't like ZoL's using procfs to expose the values, but the procfs is a straight mapping from sysctls, so that's not an issue.

Hence, I conclude that using sysctls as OpenZFS-specific identifiers is acceptable, even if they are mapped to procfs on Linux.